presto-main/src/main/java/com/facebook/presto/resourcemanager/ResourceManagerClusterStatusSender.java

```diff
@@ -0,0 +1,142 @@
+import java.util.function.Supplier;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+public class ResourceManagerClusterStatusSender
+        implements ClusterStatusSender
+{
+    private final DriftClient<ResourceManagerClient> resourceManagerClient;
+    private final InternalNodeManager internalNodeManager;
+    private final Supplier<NodeStatus> statusSupplier;
+    private final ScheduledExecutorService executor;
+    private final ResourceManagerConfig resourceManagerConfig;
+
+    private final ConcurrentMap<QueryId, PeriodicTaskExecutor> queries = new ConcurrentHashMap<>();
+
+    private PeriodicTaskExecutor nodeHeartbeatSender;
+
+    @Inject
+    public ResourceManagerClusterStatusSender(
+            @ForResourceManager DriftClient<ResourceManagerClient> resourceManagerClient,
+            InternalNodeManager internalNodeManager,
+            StatusResource statusResource,
+            @ForResourceManager ScheduledExecutorService executor,
+            ResourceManagerConfig resourceManagerConfig)
+    {
+        this.resourceManagerClient = requireNonNull(resourceManagerClient, "resourceManagerService is null");
+        this.internalNodeManager = requireNonNull(internalNodeManager, "internalNodeManager is null");
+        requireNonNull(statusResource, "statusResource is null");
+        this.statusSupplier = statusResource::getStatus;
+        this.executor = requireNonNull(executor, "executor is null");
+        this.resourceManagerConfig = requireNonNull(resourceManagerConfig, "resourceManagerConfig is null");
+    }
+
+    public ResourceManagerClusterStatusSender(
+            DriftClient<ResourceManagerClient> resourceManagerClient,
+            InternalNodeManager internalNodeManager,
+            Supplier<NodeStatus> statusResource,
+            ScheduledExecutorService executor,
+            ResourceManagerConfig resourceManagerConfig)
+    {
+        this.resourceManagerClient = requireNonNull(resourceManagerClient, "resourceManagerService is null");
+        this.internalNodeManager = requireNonNull(internalNodeManager, "internalNodeManager is null");
+        this.statusSupplier = requireNonNull(statusResource, "statusResource is null");
+        this.executor = requireNonNull(executor, "executor is null");
+        this.resourceManagerConfig = requireNonNull(resourceManagerConfig, "resourceManagerConfig is null");
+    }
+
+    @PostConstruct
+    public void init()
+    {
+        nodeHeartbeatSender = new PeriodicTaskExecutor(resourceManagerConfig.getNodeHeartbeatInterval().toMillis(), executor, this::sendNodeHeartbeat);
+    }
+
+    @PreDestroy
+    public void stop()
+    {
+        queries.values().forEach(PeriodicTaskExecutor::stop);
+        if (nodeHeartbeatSender != null) {
+            nodeHeartbeatSender.stop();
+        }
+    }
+
+    @Override
+    public void registerQuery(ManagedQueryExecution queryExecution)
```

Comment:
IIUC, this method will only be called on coordinator/dispatcher, while `sendNodeHeartbeat` method will be called on all type of nodes?

Reply:
Yes, exactly.